okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -298,6 +299,8 @@ public void streamFailed(IOException e) {
           noNewStreams = true;
           route = null;
         }
+      } else if (e instanceof ConnectionShutdownException) {
+        noNewStreams = true;
```

Comment:
I've thought about what you said, but I'm still uncertain on this. One counterexample I came up with, what if an app layer network interceptor throws an unchecked exception? Should that kill the connection? I believe the current code will preserve an HTTP/2 connection (I like that), but kill an HTTP/1.1 connection.